src/perf/event_provider.cpp

```diff
@@ -0,0 +1,389 @@
+    { "s", PERF_COUNT_HW_CACHE_RESULT_ACCESS },
+    { "-misses", PERF_COUNT_HW_CACHE_RESULT_MISS },
+};
+
+inline constexpr std::uint64_t make_cache_config(perf_hw_cache_id cache, perf_hw_cache_op_id op,
+                                                 perf_hw_cache_op_result_id op_result)
+{
+    return cache | (op << 8) | (op_result << 16);
+}
+
+template <std::size_t N, typename T>
+inline constexpr std::size_t array_size(T (&)[N])
+{
+    return N;
+}
+}
+
+namespace lo2s
+{
+namespace perf
+{
+
+static bool event_is_openable(const platform::CounterDescription& ev)
+{
+    struct perf_event_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.size = sizeof(attr);
+    attr.type = ev.type;
+    attr.config = ev.config;
+    attr.config1 = ev.config1;
+    attr.exclude_kernel = 1;
+
+    int fd = syscall(__NR_perf_event_open, &attr, 0, -1, -1, 0);
+    if (fd == -1)
+    {
+        switch (errno)
+        {
+        case ENOTSUP:
+            Log::debug() << "perf event not supported by the running kernel: " << ev.name;
+            break;
+        default:
+            Log::debug() << "perf event not available: " << ev.name;
+            break;
+        }
+        return false;
+    }
+
+    close(fd);
+    return true;
+}
+
+static void populate_event_map(EventProvider::EventMap& map)
+{
+    Log::info() << "checking available events...";
+    map.reserve(array_size(HW_EVENT_TABLE) + array_size(SW_EVENT_TABLE) +
+                array_size(CACHE_NAME_TABLE) * array_size(CACHE_OPERATION_TABLE) *
+                    array_size(CACHE_OP_RESULT_TABLE));
+    for (const auto& ev : HW_EVENT_TABLE)
+    {
+        if (event_is_openable(ev))
+        {
+            map.emplace(ev.name, ev);
+        }
+    }
+
+    for (const auto& ev : SW_EVENT_TABLE)
+    {
+        if (event_is_openable(ev))
+        {
+            map.emplace(ev.name, ev);
+        }
+    }
+
+    std::stringstream name_fmt;
+    for (const auto& cache : CACHE_NAME_TABLE)
+    {
+        for (const auto& operation : CACHE_OPERATION_TABLE)
+        {
+            for (const auto& op_result : CACHE_OP_RESULT_TABLE)
+            {
+                name_fmt.str(std::string());
+                name_fmt << cache.name << '-' << operation.name << op_result.name;
+
+                platform::CounterDescription ev(
+                    name_fmt.str(), PERF_TYPE_HW_CACHE,
+                    make_cache_config(cache.id, operation.id, op_result.id));
+
+                if (event_is_openable(ev))
+                {
+                    map.emplace(name_fmt.str(), ev);
+                }
+            }
+        }
+    }
+}
+
+static std::uint64_t parse_bitmask(const std::string& format)
+{
+    enum format_regex_groups
+    {
+        WHOLE_MATCH,
+        BIT_BEGIN,
+        BIT_END,
+    };
+
+    std::uint64_t mask = 0x0;
+
+    static const std::regex bit_mask_regex(R"((\d+)?(?:-(\d+)))");
+    static const std::sregex_iterator end;
```

Comment:
Is the std::sregex_iterator that expensive?
Otherwise I'd change it to not being static.